src/ui/BlockEditor.tsx

```diff
@@ -280,68 +243,99 @@ export type Search = {
-    props.dispatch({ type: "RESET_STORE_FOR_TESTING" });
-  }
+export type BlockEditorProps = {
+  value: string;
+  options?: Options;
+  cmOptions?: CodeMirror.EditorConfiguration;
+  keyMap?: { [index: string]: string };
+  /**
+   * id of the language being used
+   */
+  languageId: string;
+  search?: Search;
+  toolbarRef?: React.RefObject<HTMLInputElement>;
+  onBeforeChange?: IUnControlledCodeMirror["onBeforeChange"];
+  onMount: Function;
+  api?: API;
+  passedAST?: AST; // not used?
+  showDialog: Function;
+};
 
-  static defaultProps = {
-    keyMap: defaultKeyMap,
-    search: {
+const BlockEditor = (props: BlockEditorProps) => {
+  const mounted = useRef(); // from https://stackoverflow.com/a/53406363/12026982
+  const dispatch: AppDispatch = useDispatch();
+  // grab some properties from state
+  const { ast, cur, hasQuarantine } = useSelector((state: RootState) => {
+    console.log("@@@@@@@@@@@@@@@@", state);
+    return {
+      ast: state.ast,
+      cur: state.cur,
+      hasQuarantine: !!state.quarantine,
+    };
+  });
+  // NOTE(Emmanuel): this was in matchDispatchToProps, but appears unused
+  const clearFocus = () => dispatch({ type: "SET_FOCUS", focusId: null });
+
+  const {
+    options,
+    keyMap = defaultKeyMap,
+    search = {
       search: () => null,
       onSearch: () => {},
       setCursor: () => {},
       setCM: () => {},
     } as Search,
-  };
+  } = props;
+
+  let mouseUsed: boolean = false;
+  let newAST: AST = null;
+  let pendingTimeout: afterDOMUpdateHandle;
+
+  useEffect(() => {
+    if (!mounted.current) {
+      // Initial mount: create a buffer for use with copy/cut/paste
+      const clipboardBuffer = document.createElement("textarea");
+      (clipboardBuffer as HTMLTextAreaElement).ariaHidden = "true";
+      clipboardBuffer.tabIndex = -1;
+      clipboardBuffer.style.opacity = "0";
+      clipboardBuffer.style.height = "1px";
+      document.body.appendChild(SHARED.buffer);
+
+      // TODO (Emmanuel): pass SHARED fields below inside a React
+      // Context, instead of storing them in a junk drawer
+      SHARED.options = options;
+      SHARED.search = search;
+      SHARED.buffer = clipboardBuffer;
+    }
+    // refresh() after component is mounted or updated
+    pendingTimeout = setAfterDOMUpdate(refreshCM);
 
-  /**
-   * @internal
-   * Anything that didn't come from CMB itself must be speculatively
-   * checked. NOTE: this only checks the *first change* in a changeset!
-   * This is hooked up to CodeMirror's onBeforeChange event
-   */
-  handleBeforeChange = (
+    // return our cleanup function
+    return () => {
+      SHARED.buffer.remove();
+      cancelAfterDOMUpdate(pendingTimeout);
+    };
+  }, []);
+
+  // NOTE(Emmanuel): we shouldn't have to dispatch this in the constructor
+  // just for tests to pass! Figure out how to reset the store manually
+  dispatch({ type: "RESET_STORE_FOR_TESTING" });
```

Comment:
This was previously in the constructor, where it was only executed once, before mount. Not it is being executed on every render. That seems like a problem.